R/oxcAARCalibratedDate.R

```diff
@@ -206,6 +331,71 @@ plotoxcAARDateSystemGraphics <- function(x, ...){
 #' @export
 is.oxcAARCalibratedDate <- function(x) {"oxcAARCalibratedDate" %in% class(x)}
 
+#' @title Convert to a CalDates object
+#' @description Convert oxcAARCalibratedDatesList to an rcarbon CalDates object. The function is taken and adapted from rcarbon (Bevan, A. and Crema, E.R., 2019).
+#' @param x One or more calibrated dated in the form of an oxcAARCalibratedDatesList
+#' @return A CalDates object
+#' @examples
+#' \dontrun{
+#' library(oxcAAR)
+#' library(rcarbon)
+#' quickSetupOxcal()
+#' dates <- data.frame(CRA=c(3200,2100,1900), Error=c(35,40,50))
+#' rcaldates <- rcarbon::calibrate(dates$CRA, dates$Error, calCurves=rep("intcal13"))
+#' ocaldates <- oxcAAR::oxcalCalibrate(c(3200,2100,1900),c(35,40,50),c("a","b","c"))
+#' ## Convert to rcarbon format
+#' caldates.o <- as.CalDates(ocaldates)
+#' ## Comparison plot
+#' plot(rcaldates$grids[[2]]$calBP,rcaldates$grids[[2]]$PrDens,
+#' type="l", col="green", xlim=c(2300,1900))
+#' lines(caldates.o$grids[[2]]$calBP,caldates.o$grids[[2]]$PrDens, col="blue")
+#' legend("topright", legend=c("rcarbon","OxCal"), col=c("green","blue"), lwd=2)
+#' }
+#' @export
+#' @importFrom "stats" "approx"
+as.CalDates <- function(x){
+  if (!any(class(x)%in%c("oxcAARCalibratedDatesList"))){
+    stop("x must be of class oxcAARCalibratedDatesList")
+  }
+  if (any(class(x)=="oxcAARCalibratedDatesList")){
+    reslist <- vector(mode="list", length=2)
+    sublist <- vector(mode="list", length=length(x))
+    names(sublist) <- names(x)
+    names(reslist) <- c("metadata","grids")
+    ## metadata
+    df <- as.data.frame(matrix(ncol=12, nrow=length(x)), stringsAFactors=FALSE)
+    names(df) <- c("DateID","CRA","Error","Details","CalCurve","ResOffsets","ResErrors","StartBP","EndBP","Normalised","CalEPS", "Method")
+    df$DateID <- names(x)
+    df$CRA <- as.numeric(unlist(lapply(X=x, FUN=`[[`, "bp")))
+    df$Error <- as.numeric(unlist(lapply(X=x, FUN=`[[`, "std")))
+    df$CalCurve=lapply(lapply(lapply(lapply(lapply(x,FUN=`[[`,"cal_curve"),FUN=`[[`,"name"),strsplit," "),unlist),FUN=`[[`,2)
+    df$CalCurve=tolower(df$CalCurve)
+    df$ResOffsets <- NA
+    df$ResErrors <- NA
+    df$StartBP <- NA
+    df$EndBP <- NA
+    df$Normalised <- TRUE
+    df$CalEPS <- 0
+    df$Method <- 'oxcAAR'
+    reslist[["metadata"]] <- df
+    ## grids
+    for (i in 1:length(x)){
+      tmp <- x[[i]]$raw_probabilities
+      rr <- range(tmp$dates)
+      res <- 	approx(x=tmp$dates,y=tmp$probabilities,xout=ceiling(rr[1]):floor(rr[2]))
+      res$x <- abs(res$x-1950)
+      res <- data.frame(calBP=res$x,PrDens=res$y)
+      res$PrDens <- res$PrDens/sum(res$PrDens)
+      class(res) <- append(class(res),"calGrid")
+      sublist[[i]] <- res
+    }
+    reslist[["grids"]] <- sublist
+    reslist[["calmatrix"]] <- NA
+    class(reslist) <- c("CalDates",class(reslist))
+    return(reslist)
+  }
+}
```

Comment:
Not sure if this is perfectly clean.

```
> my_uncal_dates <- data.frame(bp=c(5000,4500,4000),
                             std=c(45,35,25),
                             names=c("Date 1", "Date 2", "Date 3")
                             )
> my_cal_dates <- oxcalCalibrate(my_uncal_dates$bp, my_uncal_dates$std, my_uncal_dates$names)
> as.CalDates(my_cal_dates) %>% str()
List of 3
 $ metadata :'data.frame':	3 obs. of  11 variables:
  ..$ DateID    : chr [1:3] "Date 1" "Date 2" "Date 3"
  ..$ CRA       : num [1:3] 5000 4500 4000
  ..$ Error     : num [1:3] 45 35 25
  ..$ Details   : logi [1:3] NA NA NA
  ..$ CalCurve  : chr [1:3] "data" "data" "data"
  ..$ ResOffsets: logi [1:3] NA NA NA
  ..$ ResErrors : logi [1:3] NA NA NA
  ..$ StartBP   : logi [1:3] NA NA NA
  ..$ EndBP     : logi [1:3] NA NA NA
  ..$ Normalised: logi [1:3] TRUE TRUE TRUE
  ..$ CalEPS    : num [1:3] 0 0 0
 $ grids    :List of 3
  ..$ Date 1:Classes ‘calGrid’ and 'data.frame':	545 obs. of  2 variables:
  .. ..$ calBP : num [1:545] 6009 6008 6007 6006 6005 ...
  .. ..$ PrDens: num [1:545] 0 0 0 0 0 0 0 0 0 0 ...
  ..$ Date 2:Classes ‘calGrid’ and 'data.frame':	635 obs. of  2 variables:
  .. ..$ calBP : num [1:635] 5474 5473 5472 5471 5470 ...
  .. ..$ PrDens: num [1:635] 0 0 0 0 0 ...
  ..$ Date 3:Classes ‘calGrid’ and 'data.frame':	585 obs. of  2 variables:
  .. ..$ calBP : num [1:585] 4814 4813 4812 4811 4810 ...
  .. ..$ PrDens: num [1:585] 0 0 0 0 0 ...
 $ calmatrix: logi NA
 - attr(*, "class")= chr [1:2] "CalDates" "list"
```
```
> rcarbon::calibrate(c(5000,4500,4000), errors=c(45,35,25)) %>% str()
[1] "Calibrating radiocarbon ages..."
  |=====================================================================================================| 100%
[1] "Done."
List of 3
 $ metadata :'data.frame':	3 obs. of  12 variables:
  ..$ DateID    : chr [1:3] "1" "2" "3"
  ..$ CRA       : num [1:3] 5000 4500 4000
  ..$ Error     : num [1:3] 45 35 25
  ..$ Details   : logi [1:3] NA NA NA
  ..$ CalCurve  : chr [1:3] "intcal20" "intcal20" "intcal20"
  ..$ ResOffsets: num [1:3] 0 0 0
  ..$ ResErrors : num [1:3] 0 0 0
  ..$ StartBP   : num [1:3] 55000 55000 55000
  ..$ EndBP     : num [1:3] 0 0 0
  ..$ Normalised: logi [1:3] TRUE TRUE TRUE
  ..$ F14C      : logi [1:3] FALSE FALSE FALSE
  ..$ CalEPS    : num [1:3] 1e-05 1e-05 1e-05
 $ grids    :List of 3
  ..$ 1:Classes ‘calGrid’ and 'data.frame':	336 obs. of  2 variables:
  .. ..$ calBP : num [1:336] 5922 5921 5920 5919 5918 ...
  .. ..$ PrDens: num [1:336] 1.11e-05 1.29e-05 1.49e-05 1.77e-05 2.10e-05 ...
  ..$ 2:Classes ‘calGrid’ and 'data.frame':	443 obs. of  2 variables:
  .. ..$ calBP : num [1:443] 5432 5431 5430 5429 5428 ...
  .. ..$ PrDens: num [1:443] 1.10e-05 1.27e-05 1.47e-05 1.42e-05 1.37e-05 ...
  ..$ 3:Classes ‘calGrid’ and 'data.frame':	270 obs. of  2 variables:
  .. ..$ calBP : num [1:270] 4611 4610 4609 4608 4607 ...
  .. ..$ PrDens: num [1:270] 1.19e-05 1.35e-05 1.91e-05 2.14e-05 2.14e-05 ...
 $ calmatrix: logi NA
 - attr(*, "class")= chr [1:2] "CalDates" "list"
```
There is some factual loss of information, e.g. in the `CalCurve` field.

Reply:
made an issue for the rcarbon-sync-branch